ddtrace/contrib/asgi/middleware.py

```diff
@@ -239,35 +245,29 @@ async def wrapped_send(message):
                         span.finish()
 
             async def wrapped_blocked_send(message):
-                accept_header = (
-                    "text/html"
-                    if "text/html" in _asm_request_context.get_headers().get("accept", "").lower()
-                    else "text/json"
-                )
+                status, headers, content = core.dispatch("asgi.block.started", ctx, url)[0][0]
                 if span and message.get("type") == "http.response.start":
-                    message["headers"] = [
-                        (
-                            b"content-type",
-                            b"text/html"
-                            if accept_header and "text/html" in accept_header.lower()
-                            else b"application/json",
-                        ),
-                    ]
-
-                if message.get("type") == "http.response.body":
-                    message["body"] = bytes(appsec_utils._get_blocked_template(accept_header), "utf-8")
+                    message["headers"] = headers
+                    message["status"] = status
+                elif message.get("type") == "http.response.body":
+                    message["body"] = content
                     message["more_body"] = False
-
                 try:
                     return await send(message)
                 finally:
+                    trace_utils.set_http_meta(
+                        span, self.integration_config, status_code=status, response_headers=headers
+                    )
                     if message.get("type") == "http.response.body" and span.error == 0:
                         span.finish()
 
             try:
-                if _request_blocked(span):
+                core.dispatch("asgi.start_request", "asgi")
+                if core.get_item(HTTP_REQUEST_BLOCKED, span=span):
```

Comment:
If possible, removing `span=span` from this call would be a big step in the direction of decoupling appsec from tracing. I believe it will work because of this: https://github.com/DataDog/dd-trace-py/blob/5cb0b45c7179c21f203bff2ebe4008a7b112d0a3/ddtrace/appsec/_processor.py#L269

Reply:
It's working.